src/components/tl/ucp/allgather/allgather_knomial.c

```diff
@@ -0,0 +1,210 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "config.h"
+#include "tl_ucp.h"
+#include "tl_ucp_coll.h"
+#include "tl_ucp_sendrecv.h"
+#include "core/ucc_progress_queue.h"
+#include "core/ucc_mc.h"
+#include "coll_patterns/sra_knomial.h"
+#include "utils/ucc_math.h"
+#include "utils/ucc_coll_utils.h"
+
+enum {
+    PHASE_INIT,
+    PHASE_LOOP,  /* main loop of recursive k-ing */
+    PHASE_EXTRA, /* recv from extra rank */
+    PHASE_PROXY  /* recv from extra rank */
+};
+
+#define CHECK_PHASE(_p)                         \
+    case _p:                                    \
+    goto _p;                                    \
+    break;
+
+#define GOTO_PHASE(_phase)                                                     \
+    do {                                                                       \
+        switch (_phase) {                                                      \
+            CHECK_PHASE(PHASE_EXTRA);                                          \
+            CHECK_PHASE(PHASE_LOOP);                                           \
+            CHECK_PHASE(PHASE_PROXY);                                          \
+        case PHASE_INIT:                                                       \
+            break;                                                             \
+        };                                                                     \
+    } while (0)
+
+
+#define SAVE_STATE(_phase)                      \
+    do {                                        \
+        task->reduce_scatter_kn.phase = _phase;      \
+    } while (0)
+
```

Comment:
GOTO_PHASE can't be generic since not always we need the same number of phases. E.g. there is no "PROXY" phase in reduce_scatter. If you put into common place then you will get "unused labels". Regarding moving into pattern: i don't think this is right since all those PHASEs stuff is purely tl_ucp implementation specific thing. I will try to collapse some of them in tl_ucp

Reply:
if some phases are not needed you can just create dummy label with assert so that label is not unused

Reply:
moved to common recursive_knomial.h